library/java/org/chromium/net/impl/Errors.java

```diff
@@ -0,0 +1,105 @@
+package org.chromium.net.impl;
+
+import android.util.Log;
+import androidx.annotation.StringDef;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import org.chromium.net.NetworkException;
+
+public class Errors {
+  /**Subset of errors defined in
+   * https://github.com/envoyproxy/envoy/blob/main/envoy/stream_info/stream_info.h */
+  @StringDef({EnvoyMobileError.DNS_RESOLUTION_FAILED, EnvoyMobileError.DURATION_TIMEOUT,
+              EnvoyMobileError.STREAM_IDLE_TIMEOUT, EnvoyMobileError.UPSTREAM_CONNECTION_FAILURE,
+              EnvoyMobileError.UPSTREAM_CONNECTION_TERMINATION,
+              EnvoyMobileError.UPSTREAM_REMOTE_RESET})
+  @Retention(RetentionPolicy.SOURCE)
+  public @interface EnvoyMobileError {
+    String DNS_RESOLUTION_FAILED = "0x4000000";
+    String DURATION_TIMEOUT = "0x400000";
+    String STREAM_IDLE_TIMEOUT = "0x10000";
+    String UPSTREAM_CONNECTION_FAILURE = "0x20";
+    String UPSTREAM_CONNECTION_TERMINATION = "0x40";
+    String UPSTREAM_REMOTE_RESET = "0x10";
+  }
+
+  /** Subset of errors defined in chromium/src/net/base/net_error_list.h */
+  public enum NetError {
+    ERR_NETWORK_CHANGED(-21),
+    ERR_HTTP2_PING_FAILED(-352),
+    ERR_QUIC_PROTOCOL_ERROR(-356),
+    ERR_QUIC_HANDSHAKE_FAILED(-358),
+    ERR_NAME_NOT_RESOLVED(-105),
+    ERR_INTERNET_DISCONNECTED(-106),
+    ERR_TIMED_OUT(-7),
+    ERR_CONNECTION_CLOSED(-100),
+    ERR_CONNECTION_TIMED_OUT(-118),
+    ERR_CONNECTION_REFUSED(-102),
+    ERR_CONNECTION_RESET(-101),
+    ERR_ADDRESS_UNREACHABLE(-109),
+    ERR_OTHER(-1000);
+
+    private final int value;
+
+    NetError(int errorCode) { this.value = errorCode; }
+
+    public int getValue() { return value; }
+
+    public String getModifiedString() { return "net::" + this.toString(); }
+  }
+
+  public static NetError mapEnvoyMobileErrorToNetError(long responseFlag) {
+    String errorCode = "0x" + Long.toHexString(responseFlag);
```

Comment:
Converting the long to a hex string is inefficient, can you base this on the long directly?

Reply:
By inefficient, you mean the recursive way it does this conversion right?

I've made the changes.